RavenDB-12161 Extension point to override OS metrics

haludi opened the pull request:
http://issues.hibernatingrhinos.com/issue/RavenDB-12161

haludi committed:
RavenDB-12161 Extension point for cpu usage wad added + tests for windows

haludi committed:
RavenDB-12161 Cpu usage extension point tests modified to work on Linux as well

iftahbe commented on src/Raven.Server/Config/Categories/MonitoringConfiguration.cs:
```diff
@@ -1,12 +1,23 @@
+        [ConfigurationEntry("Monitoring.Cpu.Exec", ConfigurationEntryScope.ServerWideOnly)]
+        public string CpuUsageMonitorExec { get; set; }
+
+        [Description("The command line arguments for the 'Security.MasterKey.Exec' command or executable.")]
```
forgot to change 'Security.MasterKey.Exec'

iftahbe commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+
+        private async Task ReadErrors()
+        {
+            var errors = await _process.StandardError.ReadLineAsync();
```
This will read only one line of the error.

ayende replied:
Also need error handling here

haludi replied:
By "error handling" do you mean exception from `StreamReader` method?

grisha-kotler commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -77,7 +94,7 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
             }
 
-            if (currentInfo.ActiveCores == 0)
+            if (Math.Abs(currentInfo.ActiveCores) < 0.01)
```
You need to skip those checks when using the CpuUsageExtensionPoint

haludi replied:
@grisha-kotler Why? If I will get 0 cores I want the same behavior. To return the last  Cpu usage calculated or 0

grisha-kotler commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -87,7 +104,7 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
 
             var processCpuUsage = (processorTimeDiff * 100.0) / timeDiff / currentInfo.ActiveCores;
-            if (currentInfo.ActiveCores == ProcessorInfo.ProcessorCount)
+            if ((int)currentInfo.ActiveCores == ProcessorInfo.ProcessorCount)
```
same here

maximburyak commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -18,11 +18,24 @@ public static class CpuUsage
+            get => _cpuUsageExtensionPoint;
+            set
+            {
+                _previousExtensionPointInfo = new ExtensionPointInfo(value.Data);
```
What do we do with the previous value, if we have one, and if we don't, then it should be just a Lazy value?

haludi replied:
@maximburyak The `CpuUsageExtensionPoint` should be set once in the server initialization. Maybe if I will pass it to a function it will make more sense?

maximburyak replied:
Maybe the CpuUsage class shouldn't be a "static" class, but a singleton? And this way you'll initialize all the values you need inside the constructor? It looks to me that it lacks structure and state this way..

maximburyak commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -96,6 +113,24 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
+                    Logger.Operations($"Received bad data for cpu usage from extension point : {machineCpuUsage}");
+                }
+
+                machineCpuUsage = 0;
```
Not sure that in this case we should use 0, maybe we should fall back to the "normal" calculation

haludi replied:
@maximburyak But the situation in the first place is the "normal" calculation is wrong.
In my humble opinion, it is better to display 0 that show you something is wrong

ayende replied:
0 is _not_ wrong.
0 is a valid value

haludi replied:
So should I return a negative value?

ayende replied:
Yes, and make sure that this is user visible

maximburyak commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+                Dispose();
+            }
+
+            var _ = ReadProcess(); // explicitly starting async task without waiting for it
```
What if we failed starting the process?

maximburyak commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+        {
+            while (IsDisposed == false)
+            {
+                var nextLine = _process.StandardOutput.ReadLineAsync();
```
We should Try-Catch here, we don't want an unhandled failed task, we could for an instance have a temporary out of memory problem here, that will just kill the task, without us knowing/monitoring it.

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -1628,6 +1644,7 @@ public void Dispose()
                 ea.Execute(() => _clusterMaintenanceWorker?.Dispose());
 
                 ea.ThrowIfNeeded();
+                CpuUsage.CpuUsageExtensionPoint?.Dispose();
```
This should be inside the `ea.Execute`

ayende commented on src/Raven.Server/Dashboard/ThreadsInfo.cs:
```diff
@@ -14,7 +14,7 @@ public class ThreadsInfo : AbstractDashboardNotification
         public double CpuUsage { get; set; }
 
-        public long ActiveCores { get; set; }
+        public double ActiveCores { get; set; }
```
There is no meaning from our perspective to `0.35` of a core. The extension script needs to do the transaction, not us

haludi replied:
https://github.com/ravendb/ravendb/pull/7960#discussion_r232563223

ayende commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -96,6 +113,24 @@ private static double CalculateProcessCpuUsage(ProcessInfo currentInfo, ProcessI
+        {
+            var extensionPoint = new ExtensionPointInfo(CpuUsageExtensionPoint.Data);
+            var machineCpuUsage = extensionPoint.MachineCpuUsage;
+            if (machineCpuUsage < 0)
```
In what case this can happen? And why shouldn't we abort the extension at this point?

haludi replied:
@ayende This can happen when: 
* There were issues in the extension point process and there was one try to restart
*  If the process returns a negative value.
What do you mean by abort? Just to kill the process? In the first scenario, it is already the situation. In the second scenario if I check for bottom edge maybe to check the upper side?

ayende replied:
Yes, on bad input from the script, kill it and retry.

haludi replied:
https://github.com/ravendb/ravendb/pull/7960#discussion_r232561460
And to return a negative value or keep it 0?

ayende commented on src/Raven.Server/Utils/CpuUsage.cs:
```diff
@@ -276,7 +311,25 @@ private static unsafe MacInfo GetMacInfo()
+
+            if (Logger.IsOperationsEnabled)
+            {
+                Logger.Operations($"Received bad data for active cores from extension point : {extensionPointRawData.ActiveCores}");
```
This like the admin, what is `from extension point` which one, etc?
Need to be very explicit about this.

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, string.Empty);
```
Please use a descriptive name for the tag, not `string.Empty`

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
```
use `Environment.NewLine`

ppekrol replied:
Need to use `Environment.NewLine` instead of `\n`

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
```
what resets this value?

haludi replied:
@ayende Nothing, the implementation now is if the process has an issue the system trying to restart it once and no more

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+                    _didRestart = true;
+                    NotifyWarning($"{msg} \nTherefore the process will restart.", e);
+
+                    InitNewProcessHandle();
```
what if errors are thrown here?

haludi replied:
@ayende The `InitNewProcessHandle` function just set the new process object properties and `Start` function handles itself

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                // When the process terminating, killed or exited
+            }
+            finally
```
This shouldn't be in the finally

haludi replied:
@ayende I should not catch the exception?

haludi replied:
I want to restart or dispose the object while notifying the problem any way

ayende commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            }
+        }
+
+        private void InitNewProcessHandle()
```
Why does this need to be separate from the `Start`?

ppekrol commented on src/Raven.Server/Dashboard/ThreadsInfo.cs:
```diff
@@ -14,7 +14,7 @@ public class ThreadsInfo : AbstractDashboardNotification
         public double CpuUsage { get; set; }
 
-        public long ActiveCores { get; set; }
+        public double ActiveCores { get; set; }
```
Can you explain what is the case when this is not int/long ?

ayende replied:
The problem is when running in k8s, when you can assign 0.5 cores

ayende replied:
This still needs to be changed. The translation from partial units to integer number of cores should be done in the script, not by RavenDB

haludi replied:
In a situation where the number of cores is a fraction, the process cpu calculation will be wrong. 
To calculate the process cpu usage we compare the license limit vs the actual number of cores and use the min.

ayende replied:
That is not our responsibility. That is what the script is for.
It would be doing that translation

ayende replied:
To give more context, this is important because "core" isn't well defined.
A Raspberry Pi has 4 cores, or example, but my laptop can probably beat all of them together with just one core.

The script defines what a core means in this context, not the system. 
We do ensure that we aren't going over the number of licensed core, but that is it.

iftahbe replied:
We calculate machine usage and process usage separately. If you want to delegate all that responsibility to the script, it will have to give us RavenDB's process usage as well. So the script will have to monitor the RavenDB process. It makes it much harder than just calling the K8s API metrics-server.

ayende replied:
I pretty sure that we have to. 
And we can do that in the same way we do now, just in the script

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+                }
+                else
+                {
+                    NotifyWarning($"{msg} \nTherefore the process will terminate.", e);
```
Need to use `Environment.NewLine` instead of `\n`

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+        {
+            try
+            {
+                _process.Start();
```
Wouldn't it be more convenient for the user if the script would be started every X seconds and we would only read one line from it?

ayende replied:
No, we want this to be long lived, otherwise there are a lot of extra costs.
In particular, the common scenario we see for this is making requests to a k8s controller, which means that it will need to auth, which means that we want to have connection pooling, etc.

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, string.Empty);
```
Is JSON a way to go when it comes to scripts? It is not a commonly used format.

I would expect output like:
```
80.06 3.0
```

ayende replied:
We expect this to be a full fledged program, not just a script.
And we want to make sure that we are future proof here.

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -296,6 +294,24 @@ void ConfigureKestrel(KestrelServerOptions options)
+
+            var inspector = new CpuUsageExtensionPoint(
+                _tcpContextPool,
+                Configuration.Monitoring.CpuUsageMonitorExec,
```
We probably need to escape this?

ayende replied:
Yes, we do

ppekrol commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -296,6 +294,24 @@ void ConfigureKestrel(KestrelServerOptions options)
+            var inspector = new CpuUsageExtensionPoint(
+                _tcpContextPool,
+                Configuration.Monitoring.CpuUsageMonitorExec,
+                Configuration.Monitoring.CpuUsageMonitorExecArguments,
```
We probably need to escape this?

ayende replied:
Yes

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,263 @@
+            {
+                NotifyWarning("Could not start cpu usage extension point process", e);
+
+                Dispose();
```
Need to rethrow here?

haludi replied:
@ppekrol As I understand it the error handling should be soft so I just notify there is a problem. What do you think? how the error should be handled?

ppekrol replied:
Then you probably want to return here, not 'ReadProcess' ? (check @maximburyak comment)

haludi replied:
In this case, the origin methods to calculate CPU usage give wrong values so if the process didn't start or if it failed while running I still want to read the data from the disposed object to get error value from it. Another option I can think of is to switch the `CpuUsageExtensionPoint` to "Failed object" that will do this work and use it is I can't start. Is it make sense?

ppekrol replied:
If process died, then you want to `ReadErrors`, not the normal output I think?

haludi replied:
There is a task that responsible to read errors, notify them and kill the process if needed. But after that or if the process killed/existed by another reason when for instance the Dashboard requests for CPU usage it will get negative values to indicate the problem

haludi committed:
RavenDB-12161
CpuUsageMonitorExecArguments comment fixed
CpuUsageExtensionPoint error handling
\n to Enviroment.Newline
CpuUsageExtensionPoint refactoring

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,284 @@
+                {
+                    // When the process terminating, killed or exited
+                }
+                errors = errors + await _process.StandardError.ReadToEndAsync();
```
Can't we read to end at the beginning?

haludi replied:
As I understand it the `ReadToEndAsync` would not return until the stream is closed.
So if the process send error without exiting it will continue to wait for the continuation

ppekrol commented on src/Raven.Server/Utils/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,284 @@
+            {
+                try
+                {
+                    var blittable = context.ReadForMemory(data, "cpuUsageExtensionPointData");
```
We need to dispose this

haludi committed:
RavenDB-12161 Extension point changed to recive the final process CPU usage

haludi committed:
RavenDB-12161 Refactor to CpuUsage

haludi committed:
RavenDB-12161 CPU usage file ordering

ppekrol commented:
Don't mind the random `System.Net.Sockets.SocketException : The I/O operation has been aborted because of either a thread exit or an application request` it is related to https://github.com/ravendb/ravendb/pull/8028

haludi committed:
RavenDB-12161 Dispose blittable and process, remove unused function and property

ayende commented on src/Raven.Server/Dashboard/ThreadsInfo.cs:
```diff
@@ -14,7 +14,7 @@ public class ThreadsInfo : AbstractDashboardNotification
         public double CpuUsage { get; set; }
 
-        public long ActiveCores { get; set; }
+        public double ActiveCores { get; set; }
```
This is still double, and should be int

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsage.cs:
```diff
@@ -0,0 +1,185 @@
+
+namespace Raven.Server.Utils.Cpu
+{
+    public static class CpuUsage
```
This should be an instance class, not a static class. 
The problem is that in the tests, we have multiple raven servers, and you are going to impact all of them in this way

ayende commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -1654,6 +1656,7 @@ public void Dispose()
                 ea.Execute(() => ServerMaintenanceTimer?.Dispose());
                 ea.Execute(() => AfterDisposal?.Invoke());
                 ea.Execute(() => _clusterMaintenanceWorker?.Dispose());
+                ea.Execute(CpuUsage.Dispose);
```
Only dispose the local instance, not a static one

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsage.cs:
```diff
@@ -0,0 +1,185 @@
+
+        [return: MarshalAs(UnmanagedType.Bool)]
+        [DllImport("kernel32.dll", SetLastError = true)]
+        internal static extern bool GetSystemTimes(
```
Why is this here and not in the Windows one?

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs:
```diff
@@ -0,0 +1,265 @@
+    [SuppressMessage("ReSharper", "StaticMemberInGenericType")]
+    internal abstract class CpuUsageCalculator<T> : ICpuUsageCalculator where T : ProcessInfo
+    {
+        protected static readonly Logger Logger = LoggingSource.Instance.GetLogger<MachineResources>("Server");
```
Why _do_ you have multiple static members here for each type?

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageCalculator.cs:
```diff
@@ -0,0 +1,265 @@
+
+            return new WindowsInfo
+            {
+                SystemIdleTime = CpuUsage.GetTime(systemIdleTime),
```
`CpuUsage.GetTime` move to this class

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,277 @@
+            {
+                if (IsDisposed && _logger.IsOperationsEnabled)
+                {
+                    _logger.Operations($"Try to get data from {nameof(CpuUsageExtensionPoint)} which was disposed");
```
This will be called all the time, which will flood the log

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,277 @@
+                {
+                    // When the process terminating, killed or exited
+                }
+                errors = errors + await _process.StandardError.ReadToEndAsync();
```
need new line here

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,277 @@
+                            _data.MachineCpuUsage = machineCpuUsage;
+                            _data.ProcessCpuUsage = processCpuUsage;
+                            return true;
+                        }
```
comment here that `HandleError` is called from inside the `TryGetValue`

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,277 @@
+            return false;
+        }
+
+        private bool TryGetValue(BlittableJsonReaderObject blittable, string machineCpuUsageName, out double cpuUsage)
```
`TryGetValue` is a generic name, and you are also doing error handling here, call it as such

ayende commented on src/Raven.Server/Utils/Cpu/CpuUsageExtensionPoint.cs:
```diff
@@ -0,0 +1,277 @@
+            {
+                if (_didRestart == false)
+                {
+                    _didRestart = true;
```
`_didRestart` change this to a `DateTime` and allow it if `DateTime.UtcNow - _lastRestart` is greater than 15 minutes

ayende commented on src/Raven.Server/Utils/ThreadsUsage.cs:
```diff
@@ -40,7 +41,7 @@ public ThreadsInfo Calculate()
                 threadsInfo.ActiveCores = activeCores;
 
-                if (timeDiff == 0 || activeCores == 0)
+                if (timeDiff == 0 || Math.Abs(activeCores) < 0.01)
```
activeCores should not be a double

haludi committed:
RavenDB-12161 Calculator from static to instance, terminate process condition changed, CpuUsage to CpuHelper and clean up

haludi committed:
RavenDB-12161 Test of CPU usage extension point skiped to avoid closing of socket to other tests
